File: ex/clean/src/main/kotlin/com/ktb/shop/usecases/port/ProductGateway.kt
package com.ktb.shop.usecases.port

import com.ktb.shop.entities.CategoryId
import com.ktb.shop.entities.Product
import com.ktb.shop.entities.ProductId

// OUTPUT port. With ~1K products it has both a DB-backed gateway and a cached one.
interface ProductGateway {
    fun byCategory(category: CategoryId): List<Product>
    fun byId(id: ProductId): Product?
    fun all(): List<Product>
}

File: ex/clean-go/internal/usecases/port/product_gateway.go
package port

import (
	"context"

	"github.com/ktb/clean-shop/internal/entities"
)

// ProductGateway is an OUTPUT port. With ~1K products it has both a DB-backed
// gateway and a cached one.
type ProductGateway interface {
	ByCategory(ctx context.Context, category entities.CategoryID) ([]entities.Product, error)
	ByID(ctx context.Context, id entities.ProductID) (*entities.Product, error)
	All(ctx context.Context) ([]entities.Product, error)
}
